src/java.base/share/classes/jdk/internal/foreign/layout/LayoutTransformers.java

```diff
@@ -0,0 +1,261 @@
+         */
+        static <T extends MemoryLayout>
+        LayoutTransformer<T> of(Class<T> type,
+                                Function<? super T, ? extends MemoryLayout> transformation) {
+            Objects.requireNonNull(type);
+            Objects.requireNonNull(type);
+            return LayoutTransformerImpl.of(type, transformation);
+        }
+
+        /**
+         * {@return a transformer that will remove all member names in memory layouts}
+         */
+        static LayoutTransformer<MemoryLayout> removeName() {
+            return LayoutTransformerImpl.REMOVE_NAMES;
+        }
+
+        /**
+         * {@return a transformer that will set member byte ordering to the provided
+         * {@code byteOrder}}
+         */
+        static LayoutTransformer<ValueLayout> setByteOrder(ByteOrder byteOrder) {
+            return LayoutTransformerImpl.setByteOrder(byteOrder);
+        }
+
+    }
+
+    @ValueBased
+    static final class LayoutTransformerImpl<T extends MemoryLayout>
+            implements LayoutTransformer<T> {
+
+        private final Class<T> type;
+        private final Function<? super T, ? extends MemoryLayout> op;
+
+        private LayoutTransformerImpl(Class<T> type,
+                                      Function<? super T, ? extends MemoryLayout> op) {
+            this.type = type;
+            this.op = op;
+        }
+
+        @Override
+        public MemoryLayout transform(T layout) {
+            Objects.requireNonNull(layout);
+            return op.apply(layout);
+        }
+
+        @Override
+        public MemoryLayout deepTransform(MemoryLayout layout) {
+            Objects.requireNonNull(layout);
+
+            // Breadth first
+            MemoryLayout outer = transformFlat(this, layout);
+
+            // Handle element transformation
+            return switch (outer) {
+                case SequenceLayout sl -> MemoryLayout.sequenceLayout(sl.elementCount(), deepTransform(sl.elementLayout()));
+                case StructLayout sl   -> MemoryLayout.structLayout(applyRecursively(sl));
+                case UnionLayout gl    -> MemoryLayout.unionLayout(applyRecursively(gl));
+                default -> outer;
+            };
+        }
+
+        @Override
+        public String toString() {
+            return getClass().getSimpleName() + "[" + op + "]";
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(type, op);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            return obj instanceof LayoutTransformerImpl<?> other &&
+                    type.equals(other.type) &&
+                    op.equals(other.op);
+        }
+
+        @SuppressWarnings("unchecked")
+        private static <T extends MemoryLayout> MemoryLayout transformFlat(LayoutTransformer<T> transformer, MemoryLayout l) {
+            LayoutTransformerImpl<T> t = (LayoutTransformerImpl<T>) transformer;
+            return switch (t.type) {
```

Comment:
Why the big switch here? Aren't all arms just passing the layouts to the transform? As the method is already unchecked, perhaps simplifying the code and using e.g. raw types might be a better option, as it would make the code more readable.